### PR TITLE
feat(SnQL) Add MATCH support as well as other clauses

### PR DESCRIPTION
**Pull request by evanh**
Add MATCH support for single entity queries, join queries and sub queries. Also
add support for all the smaller clauses like SAMPLE, LIMIT, OFFSET etc.

### CODE REVIEW

File: snuba/query/snql/parser.py [hunk 1]
```diff
@@ -35,17 +73,28 @@
...
+    match_clause          = space* "MATCH" space* (relationship_match+ / entity_match / subquery) space*
+    select_clause         = space* "SELECT" select_list space*
     group_by_clause       = space* "BY" group_list space*
+    where_clause          = space* "WHERE" or_expression space*
     having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
+    limit_by_clause       = space* "LIMIT" space* integer_literal space* "BY" space* column_name space*
+    limit_clause          = space* "LIMIT" space* integer_literal space*
+    offset_clause         = space* "OFFSET" space* integer_literal space*
+    sample_clause         = space* "SAMPLE" space* numeric_literal space*
```
Comment (dbennett-sentry) [inline snuba/query/snql/parser.py:92]: In the future, we will probably want to have a `SAMPLE AUTO` or `SAMPLE NONE` directive. We could fake that with `SAMPLE 0` but that might be weird.

File: snuba/query/data_source/simple.py [hunk 2]
```diff
@@ -13,6 +14,7 @@ class Entity(DataSource):
 
     key: EntityKey
     schema: ColumnSet
+    sample_rate: Optional[float] = None
```
Comment (fpacifici) [inline snuba/query/data_source/simple.py:17]: I think you would have to add this to RelationalSource as well (that's the equivalent of the entity on the physical side of the query).
Also the translator will have to be adapted to propagate this parameter.

Do you plan to remove the sample rate field from the query object as well ?

File: snuba/query/parser/expressions.py [hunk 3]
```diff
@@ -154,7 +154,7 @@ def visit_function_call(
         self,
         node: Node,
         visited_children: Tuple[
-            str, Any, List[Expression], Any, Union[Node, List[Expression]]
+            str, Any, List[Expression], Any, Union[Node, List[Expression]], Any
```
Comment (fpacifici) [inline snuba/query/parser/expressions.py:157]: What is this additional Any ?

File: snuba/query/snql/parser.py [hunk 4]
```diff
@@ -35,17 +73,28 @@
     visit_parameters_list,
     visit_quoted_literal,
 )
+from snuba.util import parse_datetime
+
 
 snql_grammar = Grammar(
     r"""
-    query_exp             = match_clause where_clause? collect_clause? group_by_clause? having_clause? order_by_clause?
+    query_exp             = match_clause select_clause group_by_clause? where_clause? having_clause? order_by_clause? limit_by_clause? limit_clause? offset_clause? sample_clause?
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:81]: Carrying over my comment from the old version of this PR

Does this mean we are finally allowing a SAMPLE cllause applied to the full query?
If yes I thinnk we need to define the semantics of a sampled joined query when the sampling key is different between the two (or more) entities that are being joined and it is not the same as the join key.

Example:
```
events sample key is event_id
groups sample key is group_id
join key is group_id
```
Clickhouse will want us to provide the sampling clause on one of the joined table or on both, the result will be different between sampling events or sampling the groups.

File: snuba/query/snql/parser.py [hunk 5]
```diff
@@ -35,17 +73,28 @@
...
 
-    match_clause          = space* "MATCH" space* open_paren clause close_paren space*
-    where_clause          = space* "WHERE" or_expression space*
-    collect_clause        = space* "COLLECT" collect_list space*
+    match_clause          = space* "MATCH" space* (relationship_match+ / entity_match / subquery) space*
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:83]: I think you should require at least one space afte the `MATCH` keyword.

File: snuba/query/snql/parser.py [hunk 6]
```diff
@@ -35,17 +73,28 @@
...
+    select_clause         = space* "SELECT" select_list space*
     group_by_clause       = space* "BY" group_list space*
+    where_clause          = space* "WHERE" or_expression space*
     having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
+    limit_by_clause       = space* "LIMIT" space* integer_literal space* "BY" space* column_name space*
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:89]: Here as well, it seems we could concatenate all these clauses without spaces. I think we should require them.
On the other hand I think we should support queries on multiple lines (which I don't think will woork with this cchange).

```
MATCH (...)
SELECT ...
```

Should be valid but I doo not see the new line anywhere in the rules.

File: snuba/query/snql/parser.py [hunk 7]
```diff
@@ -108,31 +168,52 @@ class OrTuple(NamedTuple):
...
+            args["offset"] = offset
+
+        for k, v in args.items():
+            if isinstance(v, Node):
+                args[k] = None
+
+        if args["data_source"].sample_rate is not None:
+            args["sample"] = args["data_source"].sample_rate
+
+        return Query(**args)
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:216]: Does mypy actually like this ? It should think those are all Any since visited_children is an Iterable of Any

File: snuba/query/snql/parser.py [hunk 8]
```diff
@@ -394,20 +586,85 @@ def visit_function_call(
...
-        return visit_function_call(node, visited_children)
+        name, _, params1, _, params2, alias = visited_children
+        if isinstance(alias, Node) or len(alias) == 0:
+            alias = None
+        else:
+            _, _, _, alias = alias
+            alias = alias.text
+
+        param_list1 = tuple(params1)
+        internal_f = FunctionCall(alias, name, param_list1)
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:605]: If this is a curried function the alias should be applied to the CurriedFunctionCall, not to the internal function.

File: snuba/request/schema.py [hunk 9]
```diff
@@ -112,7 +114,10 @@ def validate(self, value, dataset: Dataset, referrer: str) -> Request:
                 if key in value
             }
 
-        query = parse_query(query_body, dataset)
+        if "query" in query_body:
+            query = parse_snql_query(query_body["query"], dataset)
+        else:
+            query = parse_query(query_body, dataset)
```
Comment (fpacifici) [inline snuba/request/schema.py:120]: I think this was supposed to be in the PR that adds the schema for the SnQL query to the request schema. At this point `query_body["query"]` should never be populated.

File: tests/query/snql/test_invalid_queries.py [hunk 10]
```diff
@@ -5,28 +5,28 @@
 from snuba.query.snql.parser import parse_snql_query
 
 test_cases = [
-    # below are all cases that are not parsed completely
+    # below are case1s that are not parsed completely
```
Comment (fpacifici) [inline tests/query/snql/test_invalid_queries.py:8]: typo

File: tests/query/snql/test_query.py [hunk 11]
```diff
@@ -11,159 +15,368 @@
...
+                    "4-5",
+                    FunctionCall(None, "minus", (Literal(None, 4), Literal(None, 5))),
+                ),
+                SelectedExpression("c", Column("c", None, "c")),
+            ],
+        ),
+        id="Basic join match",
+    ),
+    pytest.param(
+        "MATCH (e: Errors) -[contains]-> (t: Transactions), (e: Errors) -[assigned]-> (ga: GroupAssignee) SELECT 4-5, c",
```
Comment (fpacifici) [inline tests/query/snql/test_query.py:52]: Do we want to accept comma or just spaces between multiple relationships ? I am ok with either but we should converge on one of the two and correct the spec accordingly.

**Review by fpacifici — CHANGES_REQUESTED** [review]

File: tests/query/snql/test_query.py (hunk 11 above)
Comment (evanh) [reply tests/query/snql/test_query.py:52]: I think commas between different relationships for readability. I'll ensure that's the case.

File: snuba/query/snql/parser.py (hunk 1 above)
Comment (evanh) [reply snuba/query/snql/parser.py:92]: I think that's a pretty simple change to make in the future, once we know for sure what keywords we want to support.

File: snuba/query/data_source/simple.py (hunk 2 above)
Comment (evanh) [reply snuba/query/data_source/simple.py:17]: I think sampling is something we should discuss to lock down how it works. I tossed this in here because I couldn't think of another way of saving the sampling rate that is passed in with each entity.

File: snuba/query/snql/parser.py (hunk 4 above)
Comment (evanh) [reply snuba/query/snql/parser.py:81]: I'll schedule something to discuss this. Probably easier than going back and forth.

File: snuba/query/snql/parser.py (hunk 7 above)
Comment (evanh) [reply snuba/query/snql/parser.py:216]: It doesn't complain for me.

File: snuba/query/snql/parser.py [hunk 12]
```diff
@@ -108,31 +185,172 @@ class OrTuple(NamedTuple):
...
+    sample_rate: Optional[float]
+
+
+class RelationshipTuple(NamedTuple):
+    lhs: EntityTuple
+    relationship: str
+    rhs: EntityTuple
+
+
+class JoinGraph(NamedTuple):
```
Comment (evanh) [inline snuba/query/snql/parser.py:200]: All the code for validating and parsing JOIN queries is here. It seems overly complex to me but I'm not sure how to simplify it down.

Comment (fpacifici) [reply snuba/query/snql/parser.py:200]: I think I have an idea about this.
I believe the root cause of the complexity lies in the graph representation that requires you to do the intermediate `_flatten` step. 
What if instead it was represented as a tree? Then creating the join clause would be achievable with a tree traversal (I think any of the traversal types will work). That acts as a reduction. The state you carre over during the traversal is the Join clause. At each step  (which is a node you are traversing) you augment it by creating a new root join between the existing join (on the left side) and the new individual node (on the right side).

Example:

Graph representation. This is actually a set of trees
```
class Edge:
     destination: Node
     conditions: # how you want to represent it

class Node:
    entity: Entity
    children: Set[Edge]

class Trees:
    roots: Set[Node]
    nodes: Set[Node]
    children: Set[Edge]
    
    def add_edge(lhs: Entity, rhs: Entity, conditions[...]
          # retrieve both rhs and lhs from the set of nodes.
          # if only lhs is there, append the edge to that node
          # if only rhs is there, add a new root and make rhs one of its children (it's better be a root otherwise the query is invalid). Remove rhs from the roots, since it is nnon more
          # if none is there, create a new root and add the child
          # if both are threre, rhs must be root. now make rhs a child of lhs and remove it from roots
```

Now if at the end this structure has multiple roots it is an invalid query.

Now you just traverse this tree recursively. At each step you return the current join clause that starts empty

```
def add_to_join(root: Node, join: Optional[JoinClause],condition) -> JoinClause
      # take care of the empty case
     join = Join(join, root, condition)
     for c in children:
         join = add_to_join(join, c, root_c_condition)
         
     return join
     # I think it may be right, then the original tree can be rebuilt by in order traversing this.
```

Of course when building the whole algorithm it will be more complex but I suspect it will still be simpler than the one in this PR since there is no need for the flattening step.

What do you think ?

**Review by fpacifici — COMMENTED** [review]
One comment inline about the join building algorithm

**Comment by evanh** [comment]
This might work. It makes sense when I think about it. Part of the flatten step is also to make it easy to reverse the list. Since the JoinClause is recursive on the RHS, the outermost JoinClause needs to have the bottom node of the tree. I'm not sure if your idea captures that but I'll poke at it.

**Comment by fpacifici** [comment]
Unless I messed up something, traversing the tree pre-order (we are adding the node itself to the join and only then processing the children by adding nodes on top of the current join clause and not by modifying the leaves) should guarantee that the root entity of the query tree is the leftmost leaf of the join tree.

(also I would advise to put that algorithms in their own module or class and unit test them thoroughly, it is much easier to prove corner cases that way than by parsing entire complex queries).

**Comment by evanh** [comment]
This now heavily conflicts with https://github.com/getsentry/snuba/pull/1480 so I'm going to close this, rebase, reformat and reopen.
